src/collider/BVHTree.cpp

```diff
@@ -158,3 +160,252 @@ BVHNode *BVHTree::AllocNode()
+
+	// Simple O(n) sort algorithm to sort all objects according to side of pivot
+	uint32_t startIdx = 0;
+	uint32_t endIdx = numKeys - 1;
+
+	// It is ~10% faster to sort the indices than to sort the whole AABB array
+	// (cache hit rate vs. memory bandwidth)
+	// Sorting in general is extremely fast.
+	while (startIdx <= endIdx && endIdx) {
+		if (keys[startIdx].center[axis] < pivot) {
+			startIdx++;
+		} else {
+			std::swap(keys[startIdx], keys[endIdx]);
+			endIdx--;
+		}
+	}
+
+	return startIdx;
+}
+
+void SingleBVHTree::ComputeOverlap(uint32_t nodeId, const AABBd &nodeAabb, std::vector<std::pair<uint32_t, uint32_t>> &out_isect) const
+{
+	PROFILE_SCOPED()
+
+	int32_t stackLevel = 0;
+	uint32_t stack[m_treeHeight + 1];
+	// Push the root node
+	stack[stackLevel++] = 0;
+
+	while (stackLevel > 0) {
+		const SingleBVHTree::Node *node = &m_nodes[stack[--stackLevel]];
+
+		if (!nodeAabb.Intersects(node->aabb))
+			continue;
+
+		if (node->kids[0] == 0) {
+			out_isect.push_back({ nodeId, node->leafIndex });
+			continue;
+		}
+
+		// Push in reverse order for pre-order traversal
+		stack[stackLevel++] = node->kids[1];
+		stack[stackLevel++] = node->kids[0];
+	}
+}
+
+void SingleBVHTree::TraceRay(const vector3d &start, const vector3d &inv_dir, double len, std::vector<uint32_t> &out_isect) const
+{
+	PROFILE_SCOPED()
+
+	int32_t stackLevel = 0;
+	uint32_t stack[m_treeHeight + 1];
+	stack[stackLevel++] = 0;
+
+	while (stackLevel > 0) {
+		uint32_t nodeIdx = stack[--stackLevel];
+		const SingleBVHTree::Node *node = &m_nodes[nodeIdx];
+
+		// Didn't intersect with the node, ignore it
+		if (!node->aabb.IntersectsRay(start, inv_dir, len))
+			continue;
+
+		// Leaf node - mark intersection and continue
+		if (node->kids[0] == 0) {
+			out_isect.push_back(node->leafIndex);
+			continue;
+		}
+
+		stack[stackLevel++] = node->kids[1];
+		stack[stackLevel++] = node->kids[0];
+	}
+}
+
+double SingleBVHTree::CalculateSAH() const
+{
+	double outSAH = 0.0;
+	auto *rootNode = GetNode(0);
+	std::vector<uint32_t> m_nodeStack = { 0 };
+
+	while (!m_nodeStack.empty()) {
+		auto *node = GetNode(m_nodeStack.back());
+		m_nodeStack.pop_back();
+
+		if (node->kids[0])
+			m_nodeStack.push_back(node->kids[0]);
+		if (node->kids[1])
+			m_nodeStack.push_back(node->kids[1]);
+
+		// surface area = 2 * width * depth * height
+		vector3d size = node->aabb.max - node->aabb.min;
+		double area = 2.f * size.x * size.y * size.z;
+
+		// Cost function according to https://users.aalto.fi/~laines9/publications/aila2013hpg_paper.pdf Eq. 1
+		outSAH += (node->kids[0] ? 1.2 : 1.0) * area;
+	}
+
+	vector3d rootSize = rootNode->aabb.max - rootNode->aabb.min;
+	float rootArea = 2.f * rootSize.x * rootSize.y * rootSize.z;
+
+	// Perform 1 / Aroot * ( SAH sums )
+	outSAH /= rootArea;
+
+	// Remove the (normalized) SAH cost of the root node from the result
+	// The SAH metric doesn't include the cost of the root node
+	outSAH -= 1.f;
```

Comment:
Similar oddness around float/double dichotomy here as above.

I also find it a little odd that our Keys and pivots are floating point vectors when we're dealing in double space too, although probably just fine,

Reply:
The keys/pivots are actually completely intended to be in single-precision compared to the actual datatypes - `SingleBVHTree::Build` is designed to try to squeeze out as much precision as possible from the mantissa by normalizing all coordinates over the total bounding volume.

This decision was made because the precision of the sort key/pivot doesn't affect the correctness of computing intersections in the final tree (that's all done in double-precision), but the cache efficiency of the sort step is very important for tree re-build performance as it's an `O(n log n)` traversal of the whole list of entries in the tree.

Reply:
Yeah, I had a feeling this was totally deliberate, interesting.

Ahh yes the whole optimize the precision of the mantissa, been ages since I thought about that - I kind of assumed that for doubles it wouldn't matter so much.  (Last time I remember thinking about this I think was when doing 7e3 math for color channels for XBox360 HDR).